src/core/model/table/column_domain.cpp

```diff
@@ -0,0 +1,302 @@
+                    partition.Insert(std::string{it.GetValue()});
+                } while (it.TryMoveToNext());
+            };
+            util::parallel_foreach(raw_domains_.begin(), raw_domains_.end(), threads_num_,
+                                   store_column);
+            block_count--;
+        }
+        return block_stream.HasNextBlock();
+    }
+
+    void RawDomainsInit(TableIndex table_id, ColumnIndex col_count) {
+        raw_domains_ = std::vector<DomainRawData>(col_count);
+        for (ColumnIndex col_id = 0; col_id != col_count; ++col_id) {
+            raw_domains_[col_id].emplace_back(table_id, col_id);
+        }
+    }
+
+    static size_t GetOptimalBlockCapacity(size_t mem_limit_bytes) {
+        static constexpr size_t kMaxCapacity = 2 << 20UL;
+        /*
+         * if the memory limit is 64 MB or more, select a block size of 2 MB
+         * otherwise, choose to divide by 32 and find the nearest power of two
+         *
+         * for example, for 16MB we choose block capacity 512 KB
+         */
+        size_t const capacity = std::pow(2, std::log2(mem_limit_bytes / 32));
+        return std::min(capacity, kMaxCapacity);
+    }
+
+public:
+    DomainManager(size_t mem_limit_mb, config::ThreadNumType threads_num)
+        : mem_limit_(mem_limit_mb << 20UL),
+          block_capacity_(GetOptimalBlockCapacity(mem_limit_)),
+          threads_num_(threads_num) {
+        RefreshMemUsage();
+    }
+
+    /// create next domains from data stream
+    void ProcessDatasetStream(TableIndex table_id,
+                              std::shared_ptr<model::IDatasetStream> const& stream) {
+        auto const col_count = static_cast<ColumnIndex>(stream->GetNumberOfColumns());
+        RawDomainsInit(table_id, col_count);
+        processed_block_count_ = 0;
+        size_t block_count = 0;
+        BlockDatasetStream block_stream{stream, block_capacity_};
+        do {
+            processed_block_count_ += block_count;
+            block_count = GetNumberOfBlocks();
+        } while (ProcessNext(block_stream, block_count));
+
+        for (DomainRawData& raw_domain : raw_domains_) {
+            if (!raw_domain.front().IsEmpty()) {
+                domains_.emplace_back(std::move(raw_domain));
+            }
+        }
+        raw_domains_.clear();
+    }
+
+    /// get processed domains
+    std::vector<ColumnDomain>&& GetDomains() && {
+        return std::move(domains_);
+    }
+};
+
+std::vector<ColumnDomain> ColumnDomain::CreateFrom(
+        std::vector<std::shared_ptr<model::IDatasetStream>> const& streams,
+        config::MemLimitMBType mem_limit_mb, config::ThreadNumType threads_num) {
+    TableIndex const table_count = static_cast<TableIndex>(streams.size());
+    DomainManager manager{mem_limit_mb, threads_num};
+    for (TableIndex table_id = 0; table_id != table_count; ++table_id) {
+        manager.ProcessDatasetStream(table_id, streams[table_id]);
+    }
+    return std::move(manager).GetDomains();
```

Comment:
Yes, we can only call this method on rvalue object, since we do not make any guarantees about the state of the manager after calling this method. Otherwise, we need to add a method to reset the manager state. But this manager is not reusable and the use of this method clearly prohibits further use of this manager